### PR TITLE
Adds initial files for sphinx/rst and readthedocs.

### PR DESCRIPTION
**Pull request by tdd11235813**
### Converts documentation to sphinx/rst for readthedocs

**Changes**
- migrates from markdown to rst formatted documentation
  - adds sphinx+readthedocs documentation
  - fixes links in documentation files
  - adds how to enable HTML doxygen target
  - adds API reference link to toctree
  - sphinx: tested html, including readthedocs, and latexpdf output
- removes markdown/
- doc/ -> docs/
- moves images folder to docs/source/ (to have absolute paths in rst)
- own logo folder for alpaka logos
- updates README.md link to logo
  - adds rtd and doxygen status badges to README.md
- Doxygen: XML output only, run script uses sed to enable HTML
  - sphinx/breathe only requires XML
- sphinx Makefile with target for link-checking

Versions: sphinx==3.0.3 & breathe==4.16.0
- Note that Breathe 4.13.0 drops support for Sphinx<2.0
- Note that Sphinx 2.0 requires Python 3.5+
- Note that breathe 4.15.0 would require Sphinx 3.0+

See **result** [here](https://alpaka.readthedocs.io/en/latest/index.html).

This PR refers to #943.

**Note** that after merging this PR, I configure readthedocs to upstream. An updated branch will trigger readthedocs/sphinx to recompile the online doc.

Sidenote: the github alpaka workflow link in README.md should refer to new alpaka-group page, I guess

Outdated:
~~Sections included:~~
~~- Introduction~~
~~- Abstraction~~
~~- Implementation~~
~~- Code Formatting~~
~~Sections not included yet:~~
~~- all the nested subsections (Thread, Block, ..., Library Interface, Mapping)~~
~~Changes~~
~~- removes doc/markdown/ and moves pics to doc/images/.~~
~~- doc/ -> docs/~~
~~- own logo folder for alpaka logos.~~
~~- updates README.md link to logo~~
~~- adds readthedocs and doxygen status badges to README.md~~

### CODE REVIEW

**Comment by SimeonEhrig** [comment]
Can you please add a section under contribution which describes how to build the sphinx doc? It is not too hard, but for somebody who has no experience with Python and Sphinx it can helps a lot of.

**Comment by BenjaminW3** [comment]
> I can continue porting

I would really like to see this!

**Review by BenjaminW3 — DISMISSED** [review]

**Comment by BenjaminW3** [comment]
How does readthedocs work?
Did you have to reserve the alpaka name? With which account did you do this? When is this documentation updated?

**Comment by tdd11235813** [comment]
> How does readthedocs work?
> Did you have to reserve the alpaka name? With which account did you do this? When is this documentation updated?

I'll add readthedocs/sphinx as section to alpaka.

The organization feature of readthedocs is only in the business version available. So the alpaka readthedocs is tied to three maintainers (@psychocoderHPC , @SimeonEhrig and me).
Please contact @psychocoderHPC if more maintainers need to be added. 

He also could remove me as maintainer for a moment, so we can try what happens, when I am out.
I created the readthedocs URL for alpaka with my account, but it does not seem to be bound on that.
I'll update it as soon as possible ;) I guess you need the missing sections, too, in order to have a complete doc, although you might want to refactor it later on. The code style guide should be reviewed, too.

**Comment by BenjaminW3** [comment]
The code style guide will mostly get removed once I finished the clang-format integration. So please do not do too much work there.

**Comment by psychocoderHPC** [comment]
@BenjaminW3 
 I can add you as maintainer  if I forger this please remember me I am currently afk.
@tdd11235813 Alpaka is open source feel free to stay readthedocs maintainer.

**Comment by tdd11235813** [comment]
ok, fine with me :) 

Info: I am about to integrate the remaining sections and I am playing with *breathe* to include *doxygen* information into sphinx. I followed mostly picongpu's workflow, i.e. doxygen Doxyfile has only XML target enabled (for breathe) and a *sed* line re-enables it in our run_doxygen.sh.
2x Doxyfiles (HTML, XML) would be too redundant, and doxygen does not offer CLI for target selection, and always creating both targets also does not seem efficient.

**Comment by tdd11235813** [comment]
update is [online](https://alpaka.readthedocs.io/en/latest/index.html), also check my [branch](https://github.com/tdd11235813/alpaka/tree/pr-readthedocs) to see the docs/ structure and the README.md update.

- `make latexpdf` produces an **error**, can you reproduce it? Seems weird, not sure how to solve it at the moment

> Syntax Error: Couldn't find trailer dictionary
> Syntax Error: Couldn't read xref table
> 
> !pdfTeX error: pdflatex (file ./alpaka.pdf): xpdf: reading PDF image failed
>  ==> Fatal error occurred, no output PDF file produced!

- `make html` shows warnings about doxygen generated things on a function, I actually want to ignore this at this stage

<details>

```
/alpaka/docs/source/usage/implementation/mapping/CUDA.rst:60: WARNING: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expecting "(" in parameters-and-qualifiers. [error at 98]
    template<typename T, std::size_t TuniqueId, typename TBlockSharedMemSt>ALPAKA_NO_HOST_ACC_WARNING ALPAKA_FN_ACC auto alpaka::block::shared::st::allocVar(TBlockSharedMemSt const & blockSharedMemSt)
    --------------------------------------------------------------------------------------------------^
```

</details>

**Review by BenjaminW3 — DISMISSED** [review]

File: docs/source/dev/sphinx.rst
Comment (SimeonEhrig) [inline docs/source/dev/sphinx.rst]: Failed, because the `Doxyfile` is in the folder `doxygen`. `doxygen doxygen/Doxyfile` works.

**Review by SimeonEhrig — COMMENTED** [review]
> update is [online](https://alpaka.readthedocs.io/en/latest/index.html), also check my [branch](https://github.com/tdd11235813/alpaka/tree/pr-readthedocs) to see the docs/ structure and the README.md update.
> 
>     * `make latexpdf` produces an **error**, can you reproduce it? Seems weird, not sure how to solve it at the moment
> 
> 
> > Syntax Error: Couldn't find trailer dictionary
> > Syntax Error: Couldn't read xref table
> > !pdfTeX error: pdflatex (file ./alpaka.pdf): xpdf: reading PDF image failed
> > ==> Fatal error occurred, no output PDF file produced!
> 
>     * `make html` shows warnings about doxygen generated things on a function, I actually want to ignore this at this stage
> 
> 
> ```
> /alpaka/docs/source/usage/implementation/mapping/CUDA.rst:60: WARNING: Error when parsing function declaration.
> If the function has no return type:
>   Error in declarator or parameters-and-qualifiers
>   Invalid C++ declaration: Expecting "(" in parameters-and-qualifiers. [error at 98]
>     template<typename T, std::size_t TuniqueId, typename TBlockSharedMemSt>ALPAKA_NO_HOST_ACC_WARNING ALPAKA_FN_ACC auto alpaka::block::shared::st::allocVar(TBlockSharedMemSt const & blockSharedMemSt)
>     --------------------------------------------------------------------------------------------------^
> ```

I can reproduce the pdf error and html warning.

**Comment by SimeonEhrig** [comment]
It may be useful to add a link to the doxygen documentation on the left side of the section overview (as with a normal section). The first time, I missed the link on the main page and did a little search.

**Comment by psychocoderHPC** [comment]
I have not followed the review but do we need to add the token to readthedocs or any other platform before we merge this PR?

File: docs/source/usage/implementation/library/rationale.rst
Comment (SimeonEhrig) [inline docs/source/usage/implementation/library/rationale.rst]: I think, here is an indentation error. `* (+) ...` and `* (-) ...` are not rendered as separate list points.
https://alpaka.readthedocs.io/en/latest/usage/implementation/library/rationale.html#implementation-variants

Same on point `* b)`

File: docs/source/dev/sphinx.rst
Comment (tdd11235813) [reply docs/source/dev/sphinx.rst]: moved Doxyfile to upper folder, which makes more sense anyway. doxygen outputs results to doxygen/ (doxygen/html, doxygen/xml, ...).
run_doxygen.sh and doc updated accordingly.

File: docs/source/usage/implementation/library/rationale.rst
Comment (tdd11235813) [reply docs/source/usage/implementation/library/rationale.rst]: thanks, fixed (click on your link again to see the result).

**Comment by tdd11235813** [comment]
- latexpdf now works
  - sphinx structures files like logo in a flat folder, so name conflict happened with the output PDF file, which had the same name alpaka.pdf like the logo (actually pdflatex included its own unfinished output file as logo^^)
  - alpaka.tex -> alpaka-doc.tex -> alpaka-doc.pdf to avoid name conflict
- Adds documentation (sphinx.rst) how to enable HTML doxygen target
- Adds API reference link to toctree

@psychocoderHPC asked for token, here I need instructions for that, what you need exactly:

> I have not followed the review but do we need to add the token to readthedocs or any other platform before we merge this PR?



Edit: rebased with develop. Going to squash commits after review process has been finished.

**Comment by SimeonEhrig** [comment]
Looks really good. Great Job!
I can confirm, that the local build of Doxygen, html and latexpdf works.

Last question: Is automatic build via CI already configured?

**Comment by tdd11235813** [comment]
> Last question: Is automatic build via CI already configured?

Note that after merging this PR, I configure readthedocs to upstream. An updated branch will trigger readthedocs/sphinx to recompile the online doc.

Squashed commits and updated starter post with changes to be committed to upstream.

**Review by SimeonEhrig — DISMISSED** [review]

**Comment by BenjaminW3** [comment]
You still have to fix a conflict in the readme before this can be merged.

**Comment by tdd11235813** [comment]
[alpaka readthedocs](https://alpaka.readthedocs.io/en/latest/) configured to upstream. Currently uses only 'latest' build, inactivated 'stable'. You can configure builds on readthedocs in "Versions".

**Comment by ax3l** [comment]
Wuiii, so cool!

@tdd11235813 @BenjaminW3 @SimeonEhrig We recently found a trick with @maxthevenet how to even include the Doxygen API HTML in versioned Sphinx HTML builds for e.g. RTD. Here is how we do it: https://github.com/openPMD/openPMD-api/pull/697

I even went one step further and created a doxygen index, which you can use with Xeus-Cling for in-jupyter `?`-help on APIs: https://openpmd-api.readthedocs.io/en/latest/details/doxygen.html
https://github.com/conda-forge/openpmd-api-feedstock/pull/46/commits/1d19a5e3cb5a4c9d67459223d51b8c4007ea94be (Current limitation: https://github.com/jupyter-xeus/xeus-cling/issues/318)

**Pull request merged**
